### PR TITLE
[SMALLFIX] Only remove blocks after journal entries are written

### PR DESCRIPTION
**Pull request by aaudiber**
This is part 1. part 2 is journaling block removal https://alluxio.atlassian.net/browse/ALLUXIO-2852

### CODE REVIEW

File: core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java [hunk 1]
```diff
@@ -1239,6 +1256,7 @@ private void deleteInternal(LockedInodePath inodePath, boolean replayed, long op
     }
 
     List<Pair<AlluxioURI, Inode>> delInodes = new LinkedList<>();
+    List<Long> delBlocks = new ArrayList<>();
```
Comment (jsimsa) [inline core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java]: nit: s/delBlocks/blocksToDelete/

File: core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java [hunk 2]
```diff
@@ -2461,47 +2479,57 @@ private void mountInternal(LockedInodePath inodePath, AlluxioURI ufsPath, long m
...
       DeleteOptions deleteOptions =
           DeleteOptions.defaults().setRecursive(true).setAlluxioOnly(true);
-      deleteAndJournal(inodePath, deleteOptions, journalContext);
+      delBlocks = deleteAndJournal(inodePath, deleteOptions, journalContext);
     } catch (DirectoryNotEmptyException e) {
       throw new RuntimeException(String.format(
           "We should never see this exception because %s should never be thrown when recursive "
-              + "is true.", e.getClass()));
+              + "is true.",
+          e.getClass()));
```
Comment (jsimsa) [inline core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java]: nit: why is this on a new line?

File: core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java [hunk 3]
```diff
@@ -2461,47 +2479,57 @@ private void mountInternal(LockedInodePath inodePath, AlluxioURI ufsPath, long m
...
    * @throws FileDoesNotExistException if the path to be mounted does not exist
    */
-  private void unmountAndJournal(LockedInodePath inodePath, JournalContext journalContext)
+  private List<Long> unmountAndJournal(LockedInodePath inodePath, JournalContext journalContext)
       throws InvalidPathException, FileDoesNotExistException, IOException {
     if (!unmountInternal(inodePath.getUri())) {
       throw new InvalidPathException("Failed to unmount " + inodePath.getUri() + ". Please ensure"
           + " the path is an existing mount point and not root.");
     }
+    List<Long> delBlocks;
```
Comment (jsimsa) [inline core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java]: nit: just `blocks` for consistency with `deleteAndJournal`

Comment (jsimsa) [inline core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java]: or use `blocksToDelete` everywhere

File: core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java [hunk 4]
```diff
@@ -2461,47 +2479,57 @@ private void mountInternal(LockedInodePath inodePath, AlluxioURI ufsPath, long m
...
     // Unmount should lock the parent to remove the child inode.
     try (JournalContext journalContext = createJournalContext();
         LockedInodePath inodePath = mInodeTree
             .lockFullInodePath(alluxioPath, InodeTree.LockMode.WRITE_PARENT)) {
       mPermissionChecker.checkParentPermission(Mode.Bits.WRITE, inodePath);
-      unmountAndJournal(inodePath, journalContext);
+      blocks = unmountAndJournal(inodePath, journalContext);
       Metrics.PATHS_UNMOUNTED.inc();
     }
+    mBlockMaster.removeBlocks(blocks, true);
```
Comment (jsimsa) [inline core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java]: `mBlockMaster.removeBlocks(blocks, true /* delete */);`

File: core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java [hunk 5]
```diff
@@ -1158,74 +1158,91 @@ public int getNumberOfPinnedFiles() {
...
+        mInodeTree.lockFullInodePath(entry.getId(), InodeTree.LockMode.WRITE)) {
+      blocks = deleteInternal(
+          inodePath, true, entry.getOpTimeMs(), DeleteOptions.defaults()
+              .setRecursive(entry.getRecursive()).setAlluxioOnly(entry.getAlluxioOnly()),
+          NoopJournalContext.INSTANCE);
+
     } catch (Exception e) {
       throw new RuntimeException(e);
     }
+    mBlockMaster.removeBlocks(blocks, true);
```
Comment (jsimsa) [inline core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java]: `mBlockMaster.removeBlocks(blocks, true /* delete */);`

File: core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java [hunk 6]
```diff
@@ -1158,74 +1158,91 @@ public int getNumberOfPinnedFiles() {
...
+    List<Long> blocks;
     Metrics.DELETE_PATHS_OPS.inc();
     try (JournalContext journalContext = createJournalContext();
          LockedInodePath inodePath = mInodeTree.lockFullInodePath(path, InodeTree.LockMode.WRITE)) {
       mPermissionChecker.checkParentPermission(Mode.Bits.WRITE, inodePath);
       mMountTable.checkUnderWritableMountPoint(path);
-      deleteAndJournal(inodePath, options, journalContext);
+      blocks = deleteAndJournal(inodePath, options, journalContext);
     }
+    mBlockMaster.removeBlocks(blocks, true);
```
Comment (jsimsa) [inline core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java]: `mBlockMaster.removeBlocks(blocks, true /* delete */);`

File: core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java (hunk 1 above)
Comment (aaudiber) [inline core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java]: done

File: core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java (hunk 2 above)
Comment (aaudiber) [inline core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java]: Intellij formats it that way with our formatter

File: core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java (hunk 3 above)
Comment (aaudiber) [inline core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java]: changed to `blocksToDelete`

File: core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java (hunk 4 above)
Comment (aaudiber) [inline core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java]: done

File: core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java (hunk 5 above)
Comment (aaudiber) [inline core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java]: done

File: core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java (hunk 6 above)
Comment (aaudiber) [inline core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java]: done

**Comment by aaudiber** [comment]
@jsimsa Comments addressed, PTAL

File: core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java [hunk 7]
```diff
@@ -1158,74 +1158,91 @@ public int getNumberOfPinnedFiles() {
...
    * @param entry the entry to use
    */
   private void deleteFromEntry(DeleteFileEntry entry) {
     Metrics.DELETE_PATHS_OPS.inc();
-    try (LockedInodePath inodePath = mInodeTree
-        .lockFullInodePath(entry.getId(), InodeTree.LockMode.WRITE)) {
-      deleteInternal(inodePath, true, entry.getOpTimeMs(),
-          DeleteOptions.defaults().setRecursive(entry.getRecursive())
-              .setAlluxioOnly(entry.getAlluxioOnly()), NoopJournalContext.INSTANCE);
+    List<Long> blocks;
```
Comment (calvinjia) [inline core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java]: Call all of these `blockToDelete` for consistency?

File: core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java [hunk 8]
```diff
@@ -1158,74 +1158,91 @@ public int getNumberOfPinnedFiles() {
...
+        mInodeTree.lockFullInodePath(entry.getId(), InodeTree.LockMode.WRITE)) {
+      blocks = deleteInternal(
+          inodePath, true, entry.getOpTimeMs(), DeleteOptions.defaults()
+              .setRecursive(entry.getRecursive()).setAlluxioOnly(entry.getAlluxioOnly()),
+          NoopJournalContext.INSTANCE);
+
     } catch (Exception e) {
       throw new RuntimeException(e);
     }
+    mBlockMaster.removeBlocks(blocks, true /* delete */);
```
Comment (calvinjia) [inline core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java]: Is this only to pick up the case where we journaled a delete file but not the corresponding delete block?

Comment (aaudiber) [inline core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java]: Yes, I think we should remove it though and avoid crossing journal boundaries during journal replay.

File: core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java (hunk 7 above)
Comment (aaudiber) [inline core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java]: done

File: core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java [hunk 9]
```diff
@@ -1158,74 +1158,89 @@ public int getNumberOfPinnedFiles() {
...
    * Writes to the journal.
    *
+   * This method does not delete blocks. Instead, it returns the blocks so that they can be deleted
+   * after the path deletion journal entry has been written. We cannot delete blocks earlier because
+   * the path deletion may fail, leaving us with files containing deleted blocks.
+   *
    * @param inodePath the path to delete
    * @param deleteOptions the method options
    * @param journalContext the journal context
+   * @return the list of deleted block ids
```
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java]: NIT: this and others, sound like the blocks were already deleted, not that they should be deleted later.

**Comment by gpang** [comment]
@aaudiber What's the difference between this and #5616 ?

**Comment by aaudiber** [comment]
Closing, review https://github.com/Alluxio/alluxio/pull/5616 instead

File: core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java (hunk 9 above)
Comment (aaudiber) [inline core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java]: Done in https://github.com/Alluxio/alluxio/pull/5616

**Pull request closed without merging**